### PR TITLE
feat: Add Toolset manifest endpoint

### PR DESCRIPTION
**Pull request by duwenxin99**
1. Calculate tool manifests when server starts.
2. Add toolset manifest endpoints.

### CODE REVIEW

File: internal/server/api.go [hunk 1]
```diff
@@ -36,8 +37,19 @@ func apiRouter(s *Server) chi.Router {
 
 func toolsetHandler(s *Server) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		toolsetName := chi.URLParam(r, "toolsetName")
-		_, _ = w.Write([]byte(fmt.Sprintf("Stub for toolset %s manifest!", toolsetName)))
+		type ToolsetManifest struct {
+			ServerVersion string `json:"serverVersion"`
+			Tools         []byte `json:"tools"`
+		}
```
Comment (kurtisvg) [inline internal/server/api.go]: This defines this struct every-time we need to use it -- any reason not to define it once place?

File: internal/server/api.go [hunk 2]
```diff
@@ -36,8 +37,19 @@ func apiRouter(s *Server) chi.Router {
 
 func toolsetHandler(s *Server) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		toolsetName := chi.URLParam(r, "toolsetName")
-		_, _ = w.Write([]byte(fmt.Sprintf("Stub for toolset %s manifest!", toolsetName)))
+		type ToolsetManifest struct {
+			ServerVersion string `json:"serverVersion"`
+			Tools         []byte `json:"tools"`
```
Comment (kurtisvg) [inline internal/server/api.go]: I think you would be better off making this an array (or map) or tools and letting them JSON themselves

File: internal/server/server.go [hunk 3]
```diff
@@ -28,13 +29,17 @@ import (
...
-	root chi.Router
+	conf    Config
+	root    chi.Router
+	version string
 
-	sources map[string]sources.Source
-	tools   map[string]tools.Tool
+	sources       map[string]sources.Source
+	tools         map[string]tools.Tool
+	toolsManifest []byte
```
Comment (kurtisvg) [inline internal/server/server.go]: nit: maybe make this "rawManifest" if it's all tools and already ready for a string

File: internal/server/api.go [hunk 4]
```diff
@@ -36,8 +37,19 @@ func apiRouter(s *Server) chi.Router {
...
+		}
+		toolsetManifest := ToolsetManifest{
+			ServerVersion: s.version,
+			Tools:         json.RawMessage(s.toolsManifest),
+		}
+		w.Header().Set("Content-Type", "application/json")
+		if err := json.NewEncoder(w).Encode(toolsetManifest); err != nil {
```
Comment (kurtisvg) [inline internal/server/api.go]: if we store bytes for the already encoded format ahead of time, we can just write the bytes here

File: internal/server/api_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	"net/http/httptest"
+	"testing"
+)
+
+func TestToolsetManifest(t *testing.T) {
+	serverVersion := "v1.2.3"
+	server := &Server{
+		version:       serverVersion,
+		toolsManifest: []byte(`{"tool1": "description"}`),
+	}
```
Comment (kurtisvg) [inline internal/server/api_test.go]: It's worthwhile to test against the server initialization logic itself in this situation

File: cmd/root.go [hunk 6]
```diff
@@ -76,6 +76,7 @@ func NewCommand() *Command {
 	}
 
 	flags := cmd.Flags()
+	flags.StringVarP(&cmd.cfg.Version, "version", "v", "1.0.0", "Server version.")
```
Comment (kurtisvg) [inline cmd/root.go]: should remove this -- will pass in as a compile time

File: internal/server/api.go [hunk 7]
```diff
@@ -15,16 +15,29 @@
...
 	"fmt"
 	"net/http"
 
 	"github.com/go-chi/chi/v5"
 	"github.com/go-chi/chi/v5/middleware"
 	"github.com/go-chi/render"
+	"github.com/googleapis/genai-toolbox/internal/tools"
 )
 
+func createToolsetMarshalJSON(s *Server) func(*tools.ToolsetConfig) ([]byte, error) {
```
Comment (kurtisvg) [inline internal/server/api.go]: should seperate into two functions: 

1. Initialize -- take ToolsetConfig convert to ToolsetManifest
2. Then can JSON Toolsetmanfiest (maybe doesn't need it's own function).

File: internal/server/api.go [hunk 8]
```diff
@@ -34,13 +47,20 @@ func apiRouter(s *Server) chi.Router {
...
+	}
+	s.manifests[""] = allToolsManifest
+	return r, nil
 }
 
 func toolsetHandler(s *Server) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		toolsetName := chi.URLParam(r, "toolsetName")
-		_, _ = w.Write([]byte(fmt.Sprintf("Stub for toolset %s manifest!", toolsetName)))
+		// toolsetName := chi.URLParam(r, "toolsetName")
```
Comment (kurtisvg) [inline internal/server/api.go]: need to update this

File: internal/server/api.go [hunk 9]
```diff
@@ -34,13 +47,20 @@ func apiRouter(s *Server) chi.Router {
 		r.Post("/", toolHandler(s))
 	})
 
-	return r
+	// Convert tool configs to JSON for manifest
+	defaultToolsetConfig := s.conf.ToolsetConfigs[""]
+	allToolsManifest, err := createToolsetMarshalJSON(s)(&defaultToolsetConfig)
```
Comment (kurtisvg) [inline internal/server/api.go]: Only needs to execute -- closure not needed

File: internal/tools/toolsets.go [hunk 10]
```diff
@@ -67,3 +67,8 @@ func (t ToolsetConfig) Initialize(toolsMap map[string]Tool) (Toolset, error) {
 	}
 	return toolset, nil
 }
+
+type ToolsetManifest struct {
+	ServerVersion string          `json:"serverVersion"`
+	ToolsManifest []*ToolManifest `json:"tools"`
```
Comment (kurtisvg) [inline internal/tools/toolsets.go]: Don't think you need a pointer here

File: internal/tools/cloud_sql_pg.go [hunk 11]
```diff
@@ -38,6 +38,10 @@ func (r CloudSQLPgGenericConfig) toolKind() string {
 	return CloudSQLPgSQLGenericKind
 }
 
+func (t CloudSQLPgGenericConfig) Describe() ToolManifest {
```
Comment (kurtisvg) [inline internal/tools/cloud_sql_pg.go]: maybe ".Manifest()"

File: internal/server/api.go (hunk 7 above)
Comment (duwenxin99) [reply internal/server/api.go]: Changed to reateToolsetManifest() which returns ToolsetManifest

File: internal/server/api.go [hunk 12]
```diff
@@ -15,16 +15,27 @@
...
 	"fmt"
 	"net/http"
 
 	"github.com/go-chi/chi/v5"
 	"github.com/go-chi/chi/v5/middleware"
 	"github.com/go-chi/render"
+	"github.com/googleapis/genai-toolbox/internal/tools"
 )
 
+func createToolsetManifest(s *Server, c tools.ToolsetConfig) tools.ToolsetManifest {
```
Comment (kurtisvg) [inline internal/server/api.go]: nit: often idiomatic to use "newToolsetManifest" for initialize functions

File: internal/server/api_test.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+					Parameters:  []tools.Parameter{{Name: "param2", Type: "type", Description: "description2"}},
+				},
+			},
+			toolsetConfigs: tools.ToolsetConfigs{
+				"toolset1": tools.ToolsetConfig{Name: "toolset1", ToolNames: []string{"tool1", "tool2"}},
+			},
+			wantStatusCode:    http.StatusOK,
+			wantServerVersion: "1.0.0",
+			wantManifests:     map[string]tools.ToolManifest{"tool1": tool1Manifest, "tool2": tool2Manifest},
+			wantErr:           false,
```
Comment (kurtisvg) [inline internal/server/api_test.go]: nit: it's OK to group "wants" like these into their own struct when you have a bunch of them to try and making it more readable

Comment (duwenxin99) [reply internal/server/api_test.go]: grouped

File: internal/server/api.go (hunk 12 above)
Comment (duwenxin99) [reply internal/server/api.go]: renamed

**Pull request merged**